fdbserver/ptxn/TLogPeekCursor.actor.cpp

```diff
@@ -0,0 +1,172 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbserver/ptxn/TLogPeekCursor.h"
+
+#include "fdbserver/ptxn/TLogInterface.h"
+#include "flow/Error.h"
+
+#include "flow/actorcompiler.h" // This must be the last #include
+
+namespace ptxn {
+
+namespace {
+
+// The deserializer will always expect the serialized data has a header. This function provides header-only serialized
+// data for the consumption of the deserializer.
+const Standalone<StringRef>& emptyCursorHeader() {
+	static Standalone<StringRef> empty;
+	if (empty.size() == 0) {
+		TeamID teamID;
+		ptxn::TLogStorageServerMessageSerializer serializer(teamID);
+		serializer.completeMessageWriting();
+		empty = serializer.getSerialized();
+	}
+	return empty;
+};
+
+struct PeekRemoteContext {
+	const Optional<UID> debugID;
+	const TeamID teamID;
+
+	// The last version being processed, the peek will request lastVersion + 1
+	Version* pLastVersion;
+
+	// THe interface to the remote TLog server
+	TLogInterfaceBase* pTLogInterface;
+
+	// Deserializer
+	TLogStorageServerMessageDeserializer* pDeserializer;
+
+	// Deserializer iterator
+	TLogStorageServerMessageDeserializer::iterator* pDeserializerIterator;
+
+	// If not null, attach the arena in the reply to this arena, so the mutations will still be
+	// accessible even the deserializer gets destructed.
+	Arena* pAttachArena;
+
+	PeekRemoteContext(const Optional<UID>& debugID_,
+	                  const TeamID& teamID_,
+	                  Version* pLastVersion_,
+	                  TLogInterfaceBase* pInterface_,
+	                  TLogStorageServerMessageDeserializer* pDeserializer_,
+	                  TLogStorageServerMessageDeserializer::iterator* pDeserializerIterator_,
+	                  Arena* pAttachArena_ = nullptr)
+	  : debugID(debugID_), teamID(teamID_), pLastVersion(pLastVersion_), pTLogInterface(pInterface_),
+	    pDeserializer(pDeserializer_), pDeserializerIterator(pDeserializerIterator_), pAttachArena(pAttachArena_) {
+
+		ASSERT(pTLogInterface != nullptr && pDeserializer != nullptr && pDeserializerIterator != nullptr);
+	}
+};
+
+ACTOR Future<bool> peekRemote(PeekRemoteContext peekRemoteContext) {
+	state TLogPeekRequest request;
+
+	request.debugID = peekRemoteContext.debugID;
+	request.beginVersion = *peekRemoteContext.pLastVersion + 1;
+	request.endVersion = -1; // we *ALWAYS* try to extract *ALL* data
+	request.teamID = peekRemoteContext.teamID;
+
+	try {
+		state TLogPeekReply reply = wait(peekRemoteContext.pTLogInterface->peek.getReply(request));
```

Comment:
Related to the above comment. When we have more than one interfaces, we can randomly choose one interface or use `loadBalance()` to peek data.

If you choose a random one, and add a comment that the (better) alternative is to use `loadBalance()`.